firebase-firestore/src/test/java/com/google/firebase/firestore/model/MutationSquashTest.java

```diff
@@ -0,0 +1,449 @@
+    if (mutations.isEmpty()) {
+      List<List<E>> result = Lists.newArrayList();
+      result.add(Lists.newArrayList());
+      return result;
+    }
+
+    E first = mutations.remove(0);
+    List<List<E>> returnValue = Lists.newArrayList();
+    // Generate permutation with first element removed
+    List<List<E>> permutations = generatePermutations(mutations);
+    for (List<E> permutation : permutations) {
+      // Insert `first` to each possible position of the permutation.
+      for (int index = 0; index <= permutation.size(); index++) {
+        List<E> permutationCopy = Lists.newArrayList(permutation);
+        permutationCopy.add(index, first);
+        returnValue.add(permutationCopy);
+      }
+    }
+    return returnValue;
+  }
+
+  private void squashRoundTrips(MutableDocument doc, Mutation... mutations) {
+    MutableDocument toApplySquashedMutation = doc.clone();
+    Timestamp now = Timestamp.now();
+
+    FieldMask mask = FieldMask.emptyMask();
+    for (Mutation m : mutations) {
+      mask = m.applyToLocalView(doc, now, mask);
+    }
+
+    Mutation squashed = getSquashedMutation(doc, mask);
+
+    if (squashed != null) {
+      squashed.applyToLocalView(toApplySquashedMutation, now, FieldMask.emptyMask());
+    }
+
+    assertEquals(doc, toApplySquashedMutation);
+  }
+
+  @Nullable
+  // TODO(Overlay): This is production code, find a place for this.
+  private Mutation getSquashedMutation(MutableDocument doc, FieldMask mask) {
+    Mutation squashed = null;
+    if (mask.isAllFields()) {
+      if (doc.isNoDocument()) {
+        squashed = new DeleteMutation(doc.getKey(), Precondition.NONE);
+      } else {
+        squashed = new SetMutation(doc.getKey(), doc.getData(), Precondition.NONE);
+      }
+    } else if (mask.isByMask()) {
+      ObjectValue docValue = doc.getData();
+      ObjectValue objectValue = new ObjectValue();
+      HashSet<FieldPath> maskSet = new HashSet<>();
+      for (FieldPath path : mask.getMask()) {
+        if (!maskSet.contains(path)) {
+          Value value = docValue.get(path);
+          // If it is deleting a nested field, we take the immediate parent as the mask used to
+          // construct resulting mutation.
+          // Justification: Nested fields can create parent fields implicitly, if they are
+          // deleted in later mutations, there is no trace that the empty parent fields should
+          // still remain (which is expected).
+          // Consider mutation (foo.bar 1), then mutation (foo.bar delete()).
+          // This leaves the final result (foo, {}). Despite `doc` has the correct result,
+          // `foo` is not in `mask`, and the resulting mutation will miss `foo` completely without
+          // this logic.
+          if (value == null && path.length() > 1) {
+            path = path.popLast();
+          }
+          objectValue.set(path, docValue.get(path));
+          maskSet.add(path);
```

Comment:
This code snippet should be unit tested explicitly once moved to production code.

I also just figured out why my original proposal was working: `ObjectValue.set('a.b', delete())` creates field 'a', which it should not, but it made my original implementation work.

Reply:
Hm, I think the problem that caused my test to fail was that the both the original and the new doc did not have an entry for "bar". Not sure how this happened, but this points to another thing we need to do - we should convert most test to compare against a hard-coded end state rather than a computed end state.